Add more about the calculation properties (#98)
port-labs/port-docs: docs/platform-overview/port-components/calculation-properties.md

Calculation Properties allow you to use existing properties defined on [Blueprints](./blueprint), either directly or by using [Relations](./relation) and [Mirror Properties](./mirror-properties), in order to create new properties by using the [`jq`](https://github.com/stedolan/jq) processor for `JSON`.

Calculation Properties make it easier to define properties that are based on values from other properties, with the added ability to transform the data.

The Calculation Properties support the types `string`, `number`, `object`, `array`, `boolean`, `yaml` .

With Calculation Properties you can:

- Filter/Select/Slice/Concat from an existing property.
- Create math equations or modification. For example - calculate required disk storage by specifying page size, and number of pages needed.

> - Merge complex properties, including deep-merge and overriding.

Author: On purpose,
(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/concat)

Here is a few exmaple how Calcualtion property can be beneficial:

### Concat Strings

Assume you have properties called `str1` with the value `hello` and `str2` of type `string` with the value `world`.
This is the formula get `hello world`

```json showLineNumbers
{
  "title": "Concat strings example",
  "type": "string",
  "calculation": ".properties.str1 + .properties.str2"
}
```

:::tip
You can concat properties to a default string, For example: '"https://" + .properties.str1'
:::

### Slice Array

Assume you have properties called `array1` of type `array` with the value `[1,2,3,4]`. This is the formula to get the result `[2,3,4]`

```json showLineNumbers
{
  "title": "Slice array example",
  "type": "string",
  "calculation": ".properties.array1[1:4]"
}
```

### Merge Objects

Assume you have properties called `deployed_config` with the value `{cpu: 200}` and `service_config` with the value `{memory: 400}`,
and we want to merge between these objects:

```json showLineNumbers
"calculationProperties" : {
    "merge_config": {
        "title": "Merge config",
        "type": "objet",
        "calculation": ".properties.deployed_config * .properties.service_config",
    }
}
```

> The result will be `{cpu: 200, memory: 400}`.

Reviewer: Add an example of merges where the actual order has meaning (see the note block I added below), and show those two cases (similar to the example I sent you in slack earlier)

**Notice**: This is a deep merge, The last property will **take presents** if there is any conficts of properties.

:::tip

> For [Yaml](./blueprint.md#yaml) properties, the synatx of the calculation will remain the same, but the Type will be `string`

Reviewer: ```suggestion
For [Yaml](./blueprint.md#yaml) properties, the synatx of the calculation will remain the same, but the `type` will be `string`
```

> and the format will be `yaml`.

Reviewer: ```suggestion
and the `format` will be `yaml`.
```

```json showLineNumbers

"properties":{
    "MBMemory":{
        "type": "number"
    },
},
"calculationProperties" : {
    "GBMemory": {
        "title": "GB Memory",
        "type": "number",
        "calculation": ".properties.MBMemory / 1024"
    }
}
```

In this instance, if you have a `MBMemory` **object** property with the value:

```json showLineNumbers
{
  "MBMemory": 2048
}
```

Then the `GBMemory` Calculation Property value will be:

```json showLineNumbers
{
  "GBMemory": 2
}
```

## Examples

Here are a few examples for usage of Calculation properties:

### Concat Strings

Assume you have two `string` properties: one is called `str1` with the value `hello`, the other is called `str2` with the value `world`.
The following calculation will result `hello world`:

```json showLineNumbers
{
  "title": "Concat strings example",
  "type": "string",
  "calculation": ".properties.str1 + .properties.str2"
}
```

:::tip
If you want to provide your own string template to concat properties with, you can do so by wrapping your template string with single quotes (`'`), for example: `'https://' + .properties.str1'`
:::

> ### Slice Array

Reviewer 2: ```suggestion
### Slice array
```

Assume you have an `array` property called `array1` with the value `[1,2,3,4]`. You can use the following slicing calculation to get the result `[2,3,4]`:

```json showLineNumbers
{
  "title": "Slice array example",
  "type": "string",
  "calculation": ".properties.array1[1:4]"
}
```

> ### Merge Objects
> 
> Assume you have two `object` properties: one called `deployed_config` with the value `{cpu: 200}`, the other called `service_config` with the value `{memory: 400}`,and You can merge these two object properties and get a unified config by using the following calculation:

Reviewer 2: ```suggestion
### Merge objects

Assume you have two `object` properties: one called `deployed_config` with the value `{cpu: 200}`, the other called `service_config` with the value `{memory: 400}`. You can merge these two object properties and receive a unified config by using the following calculation:
```

```json showLineNumbers
"calculationProperties" : {
    "merge_config": {
        "title": "Merge config",
        "type": "objet",
        "calculation": ".properties.deployed_config * .properties.service_config",
    }
}
```

The result will be `{cpu: 200, memory: 400}`.

> :::info Object merging
> 
> - Object merging performs a deep merge, nested keys from the original objects will also appear in the resulting merged object.
> - If the same `key` appears in one or more of the merged properties, the last property that appears will have its `keys` take precedence over the `keys` of properties that appeared earlier in the calculation.
> 
> For example, Let's assume we have 2 properties with type `object`, and we want to perform a deep merge between them:

Reviewer 2: ```suggestion
:::Info object merging

- Object merging performs a deep merge, resulting in nested keys from the original objects appearing in the resulting merged object.
- If the same `key` appears in one or more of the merged properties, the last property that appears will have its `keys` take precedence over the `keys` of properties that appeared earlier in the calculation.

For example, Let's assume we have 2 properties with type `object`, and we want to perform a deep merge between them:
```

```json showLineNumbers
{
  "obj1": {
    "cpu": 200
  },
  "obj2": {
    "cpu": 400
  }
}
```

If the calculation is `".properties.obj1 * .properties.obj2"` , the result will be `{cpu: 400}`,

If the calculation is `".properties.obj2 * .properties.obj1"` , the result will be `{cpu: 200}`,

For merging YAML properties, the merging behavior will be the same, but if you specify `type: "string` and `format: "yaml"`, the result will be a YAML object.
:::

> ## Calculation property Edge cases
> 
> In some occasions, if the key contains special characters or starts with a digit, you need to surround it with double quotes like this: .`"foo$"`
> For example, if you want to use your `on-call` property in a Calculation property (note the single quotes (`'`) around `on-call`):

Reviewer 2: ```suggestion
## Calculation Property edge cases

Sometimes, if the key contains special characters or starts with a digit, you need to surround it with double quotes like this: `"foo$"`.
For example, if you want to use your `on-call` property in a Calculation Property (note the single quotes (`'`) around `on-call`):
```
